### ISSUE TITLE
Add a way to specify beamer frame subtitles in pandoc markdown

### ISSUE THREAD

**Issue by YaLTeR** [jgm/pandoc#5031]
`pandoc 2.3.1`
Currently pandoc uses markdown headings as frame titles and allows specifying frame options via heading attributes like this:
```markdown
# This is a frame title {.allowframebreaks}
```
which gets turned into:
```tex
\begin{frame}[allowframebreaks]{This is a frame title}
```
However, frames also support subtitles, either via an additional argument:
```tex
\begin{frame}[options]{title}{subtitle}
```
or via a special command:
```tex
\begin{frame}[options]{title}
  \framesubtitle{subtitle}
```

Currently the only way of specifying the subtitle seems to be inserting the literal tex `\framesubtitle{subtitle}` after a heading, which is sub-ideal (I imagine other presentation formats also support slide subtitles).

I'm not familiar with pandoc's architecture so I'm not sure how the subtitles could be represented in the markdown input. Perhaps with a flag to specify that the last level of headings goes into frame subtitles?

**Comment by krishnakumarg1984** [comment]
Any updates on this?

**Comment by a-berg** [comment]
still relevant to this day.

**Comment by drupol** [comment]
I'm also looking for this.

**Comment by mhwombat** [comment]
Also looking for this.

**Comment by jgm** [comment]
Currently,

>   * Headings *below* the slide level in the hierarchy create
>    headings *within* a slide.  (In beamer, a "block" will be
>   created.  If the heading has the class `example`, an
>   `exampleblock` environment will be used; if it has the class
>   `alert`, an `alertblock` will be used; otherwise a regular
 >  `block` will be used.)

I suppose we could change this so that, if there is just one level N+1 heading after the level N slide heading, and no other level N+1 headings in the slide, the level N+1 heading is treated as a subtitle.  That would allow:

```
# My slide

## The subtitle

The content

# Next slide
```

**Comment by drupol** [comment]
Since the subtitle of a LaTeX frame can be set as such

```latex
\begin{frame}{title}{subtitle}

\end{frame}
```

Maybe we could image to have this? 

```markdown
# title{subtitle}
```

or 

```markdown
#{subtitle} title
```

?

**Comment by jgm** [comment]
Do you have objections to the proposal I made? It seems much more natural than introducing a new `{}` syntax.

**Comment by drupol** [comment]
> Do you have objections to the proposal I made? It seems much more natural than introducing a new `{}` syntax.

If we use `## subtitle`, what is going to replace the original `## section` then ?

**Comment by jgm** [comment]
Here slide level is #.  So a ## heading would normally produce a beamer "block."  The proposal is that, in the specific case where the slide starts with the ## heading and there are no other sub-headings, it be interpreted as a subtitle.  The thought is that there isn't much use for a single block on one slide.

**Comment by jgm** [comment]
If you did want a block, you could e.g. insert a blank HTML comment between the # and the ##.

Another possibility would be to require that the ## have class `subtitle`, though we usually try to avoid English-word syntax.

**Comment by drupol** [comment]
I did some tests,

Here's the MD file:

<details>

```markdown
---
title: "My wonderful presentation"
author: "Author name"
institute: "Institute name"
date: date
documentclass: beamer
beamer: true
theme: "ec"
notes: false
---

# Presentation title

## Frame title

### Frame subtitle

Test

# Presentation title 2

## Frame title 2

### Frame subtitle 2

Test
```

</details>

If you want to render it, you can run the build with Nix with one single command:

```shell
nix run git+https://code.europa.eu/pol/european-commission-latex-beamer-theme/#pandoc-presentation -- test.md
open pandoc-presentation.pdf
```

Basically, I wish I could keep this behavior, but with subtitles on frames.

I'm afraid that what you're proposing is going to make it not work like that anymore.

Honestly, your proposal is perfectly fine, it would be nice to at least make it work. Maybe in the end I'm wrong and everything will be fine!

**Comment by jgm** [comment]
I'm not sure what you're trying to do here.  In your example the slide level would be 3, so this isn't the sort of structure that would get subtitles under my proposal.

**Comment by drupol** [comment]
I made further tests and I think I should use `--slide-level 3` instead of 2 once you made your changes in Pandoc.

As such:

`# slide category`

`## Frame title`

`### Frame subtitle`

Here's two render of the markdown file, one using `--slide-level 3` ([pandoc-presentation-l2.pdf](https://github.com/jgm/pandoc/files/10414538/pandoc-presentation-l2.pdf)), the other using `--slide-level 2` ([pandoc-presentation-l3.pdf](https://github.com/jgm/pandoc/files/10414539/pandoc-presentation-l3.pdf))

**Comment by jgm** [comment]
No, you'd use slide-level 2.  The slide level is the level of the slide headings.

**Comment by jgm** [comment]
But I'm not sure what you think your tests are showing: I haven't made the changes yet, so this just displays current pandoc behavior.  With the proposed change, the "Frame subtitle" would show up as a proper subtitle rather than the heading of a block.

**Comment by drupol** [comment]
Allright. I'm looking forward to have support of subtitles in Pandoc :) I'll provide feedback asap.

**Comment by lucasreddinger** [comment]
This seems like a nice implementation to me.

I look forward to it! Thanks for your work on this!

Best,
Lucas

**Comment by b3** [comment]
> Another possibility would be to require that the ## have class `subtitle`, though we usually try to avoid English-word syntax.

I like this proposition and use it on my files with the following simple lua filter.

```lua
iif FORMAT:match 'beamer' then
  function Header (elem)
     if elem.level == (PANDOC_WRITER_OPTIONS.slide_level + 1) and
        elem.classes:includes("subtitle", 1) then
        return {
           pandoc.RawInline('latex', '\\framesubtitle{'),
           elem.content,
           pandoc.RawInline('latex', '}')
        }
     else
        return elem
     end
  end
end
```

It kinda works.

I am far from an expert in pandoc inside mechanisms as well as in lua and do not yet understand why my lua code produce something like that in LaTeX (why is there newlines around the content):

```latex
\framesubtitle{                                                                 
                                                                                
 This one has a subtitle                                                         
                                                                                 
 }
```

**Comment by drupol** [comment]
Hello,

Any news on this feature?

**Comment by mbanck-cd** [comment]
> > Another possibility would be to require that the ## have class `subtitle`, though we usually try to avoid English-word syntax.
> 
> I like this proposition and use it on my files with the following simple lua filter.
> 
> if FORMAT:match 'beamer' then
>   function Header (elem)
>      if elem.level == (PANDOC_WRITER_OPTIONS.slide_level + 1) and
>         elem.classes:includes("subtitle", 1) then
>         return { pandoc.RawInline('latex',
>                                   '\\framesubtitle{'
>                                   .. pandoc.utils.stringify(elem.content)
>                                   .. '}') }
>      else
>         return elem
>      end
>   end
> end

That's neat, but it does not seem to work with markup inside the subtitle, e.g.:
```
## Title
### \texttt{sub\_title} { .subtitle }
```
Where specifying the LaTeX directly works (but breaks when using po4a for translation):
```
## Title
\framesubtitle{\texttt{sub\_title}}
```
Maybe I am abusing pandoc here, though, and frame(sub)titles are never meant to be rendered like this?

**Comment by jgm** [comment]
> it does not seem to work with markup inside the subtitle

Use markdown there, not raw LaTeX.

And then avoid `stringify` which will get rid of formatting.  Try this:

```lua
 { pandoc.RawInline('latex', '\frame-subtitle'),
   pandoc.Span(elem.content) }
```

**Comment by mbanck-cd** [comment]
Ok thanks, that got me further, but it produces (if I add an additional backslash for escaping) empty newlines so that LaTeX does not consider it an argument and renders the subtitle as regular text on the frame:
```
\begin{frame}[fragile]{Title}
\protect\hypertarget{title}{}
\framesubtitle

{\texttt{sub\_title}}
```

It seems this works a bit better, i.e. wrapping in `{ }`:

```
        return { pandoc.RawInline('latex', '\\framesubtitle{'),
                 pandoc.Span(elem.content),
                 pandoc.RawInline('latex', '}') }
```

It produces this LaTeX:

```
\begin{frame}[fragile]{Title}
\protect\hypertarget{title}{}
\framesubtitle{

{\texttt{sub\_title}}

}
```
out of this markdown (at `--slide-level=2`)
```
## Title
### `sub_title` { .subtitle }
```

However, the newlines still lead to the subtitle being overly large in beamer's PDF, and if I remove them the subtitle is indented because it is on a new line; the only LaTeX that really works is everything on one line line (`\framesubtitle{{\texttt{sub\_title}}}`) is that possible with the lua filter?

**Comment by jgm** [comment]
I don't understand why we're getting these blank lines. That shouldn't happen.

**Comment by jgm** [comment]
Oh, it's because we're returning a list of inlines to replace a block element (Header).
 You want:
```lua
if FORMAT:match 'beamer' then
  function Header (elem)
     if elem.level == (PANDOC_WRITER_OPTIONS.slide_level + 1) and
        elem.classes:includes("subtitle", 1) then
        return pandoc.Plain{
          pandoc.RawInline('latex', '\\framesubtitle'),
          pandoc.Span{pandoc.Str "hi"}
        }
     end
  end
end
```

**Comment by mbanck-cd** [comment]
Thanks a lot, that works (using `pandoc.Span(elem.content)`).

~~Unfortunately, `po4a`'s markdown parser/filter is not smart enough to keep the line wrapping intact~~
Actually, if you seperate the title and subtitle with a newline, it works, so:
```
## Title

### `sub_title` { .subtitle }
```
